lua/telescope/sorters.lua

```diff
@@ -70,13 +71,229 @@ sorters.get_levenshtein_sorter = function()
+          cc / #areas[curr.area_idx].word + -- More percent of an area matched
+          (curr.exact_match and 0.1 or 0) + -- case matching
+          (curr.area_idx == area_count and 0.5 or 0)) -- A bonus for final word
+
+        prev_area = curr.area_idx
+        prev_offset = curr.area_offset
+      end
+
+      if out < score then
+        out = score
+      end
+    end
+
+    return out
+  end
+
+  local function find_points(areas, start_points, prompt, prompt_idx, exact_match)
+
+    local local_points = {}
+    local start_point = start_points[1]
+    local area_idx = start_point.area_idx
+    local area_offset = start_point.area_offset
+
+    repeat
+      local curr_area = areas[area_idx]
+      local target = prompt:byte(prompt_idx)
+
+      if target == os_sep then
+        -- I HATE THIS WTHTY ONTHOEUN YROEU NTOHEUNCFINTHINTHSN ONTTDH E MT
+        return start_points
+      end
+
+      for idx = area_offset + 1, #curr_area.word do
+        if target == curr_area.word:byte(idx) or
+          not exact_match and target - 32 == curr_area.word:byte(idx) then
+          table.insert(local_points, {
+              area_idx = area_idx,
+              area_offset = idx,
+              exact_match = target == curr_area.word:byte(idx),
+              is_upper_case = target >= 65 and target <= 90,
+            })
+        end
+      end
+
+      area_idx = area_idx + 1
+      area_offset = 0
+
+    until area_idx > #areas
+
+    local return_points = {}
+    for idx = 1, #start_points do
+      local point = start_points[idx]
+      for local_idx = 1, #local_points do
+        local local_point = local_points[local_idx]
+        if local_point.area_idx == point.area_idx and
+          local_point.area_offset > point.area_offset or
+          local_point.area_idx > point.area_idx then
+          table.insert(return_points, {
+              area_idx = local_point.area_idx,
+              area_offset = local_point.area_offset,
+              previous = point,
+              exact_match = local_point.exact_match,
+              is_upper_case = local_point.is_upper_case,
+            })
+        end
+      end
+    end
+
+    return return_points
+  end
+
+  local line_cache = {}
+  local function get_cache_for(line, prompt)
+    if not line_cache[line] then
+      line_cache[line] = Cache.PromptCache:new()
+    end
+
+    return line_cache[line]:get_cache(prompt)
+  end
+
+  local function cache_line_results(line, prompt, results)
+    local cache = line_cache[line]
+    cache:set_cache(prompt, results)
+  end
+
+  local function cache_remove_line(line)
+    line_cache[line] = nil
+  end
+
+  return Sorter:new {
+    scoring_function = function(_, prompt, line)
+      if prompt == "" then
+        return 1
+      end
+
+      local cache = get_cache_for(line, prompt) or {}
+      local areas = cache.areas or create_areas(line)
+      local new_cache = {
+        areas = areas
+      }
```

Comment:
I'm not sure why you can't use a default_dict strategy here. It seems like it would provide you the same idea (at least for areas) with less overhead.